cpp/daal/src/algorithms/tsne/tsne_gradient_descent_fpt.cpp

```diff
@@ -0,0 +1,977 @@
+        for (IdxType i = 0; i < 4; i++)
+        {
+            const auto ch = child[k * 4 + i];
+            if (ch >= 0)
+            {
+                if (i != j)
+                {
+                    // move children to front (needed later for speed)
+                    child[k * 4 + i] = -1;
+                    child[k * 4 + j] = ch;
+                }
+                if (ch >= N)
+                {
+                    // child is a cell
+                    start[ch] = begin;
+                    begin += count[ch]; // add #bodies in subtree
+                }
+                else if (begin <= nNodes && begin >= 0)
+                {
+                    // child is a body
+                    sort[begin++] = ch;
+                }
+                j++;
+            }
+        }
+        k -= dec; // move on to next cell
+    }
+    return services::Status();
+}
+
+template <typename IdxType, typename DataType, daal::CpuType cpu>
+services::Status repulsionKernelImpl(const DataType theta, const DataType eps, const IdxType * sort, const IdxType * child, const DataType * mass,
+                                     const DataType * posx, const DataType * posy, DataType * repx, DataType * repy, DataType & zNorm,
+                                     const IdxType nNodes, const IdxType N, const DataType & radius, const IdxType & maxDepth)
+{
+    SafeStatus safeStat;
+
+    //struct for tls
+    struct RepulsionTask
+    {
+    public:
+        DAAL_NEW_DELETE();
+        DataType * sumData;
+        IdxType * posData;
+        IdxType * nodeData;
+
+        static RepulsionTask * create(const IdxType maxDepth)
+        {
+            auto object = new RepulsionTask(maxDepth);
+            if (object && object->isValid()) return object;
+            delete object;
+            return nullptr;
+        }
+
+        bool isValid() const { return _sum.get() && _pos.get() && _node.get(); }
+
+    private:
+        RepulsionTask(IdxType maxDepth)
+        {
+            _sum.reset(1);
+            sumData = _sum.get();
+            services::internal::service_memset_seq<DataType, cpu>(sumData, 0, 1);
+
+            _pos.reset(maxDepth);
+            posData = _pos.get();
+            services::internal::service_memset_seq<IdxType, cpu>(posData, 0, maxDepth);
+
+            _node.reset(maxDepth);
+            nodeData = _node.get();
+            services::internal::service_memset_seq<IdxType, cpu>(nodeData, 0, maxDepth);
+        }
+
+        TArrayScalable<DataType, cpu> _sum;
+        TArrayScalable<IdxType, cpu> _pos;
+        TArrayScalable<IdxType, cpu> _node;
+    };
+
+    //initialize arrays
+    services::internal::service_memset<DataType, cpu>(repx, 0., nNodes + 1);
+    services::internal::service_memset<DataType, cpu>(repy, 0., nNodes + 1);
+    zNorm = 0.;
+
+    const IdxType fourNNodes      = 4 * nNodes;
+    const DataType thetaSquared   = theta * theta;
+    const DataType radiusdSquared = radius * radius;
```

Comment:
```suggestion
    const DataType radiusSquared = radius * radius;
```
Is this a misspelling?